### PR TITLE
damask-3.0.0-alpha5

### PR DESCRIPTION
**Pull request by MarDiehl**
update from 3.0.0-alpha4 to damask-3.0.0-alpha5.
supports PETSc 3.16.x

### CODE REVIEW

File: var/spack/repos/builtin/packages/damask-grid/PETSc-HDF5-Fortran.patch [hunk 1]
```diff
@@ -0,0 +1,10 @@
+--- damask-3.0.0-alpha5-orig/CMakeLists.txt	2021-10-07 11:15:35.325869586 +0200
++++ damask-3.0.0-alpha5/CMakeLists.txt	2021-10-07 11:19:19.081592374 +0200
+@@ -90,6 +90,7 @@
+ string (REGEX MATCHALL "-[lLW]([^\" ]+)" PETSC_EXTERNAL_LIB "${PETSC_EXTERNAL_LIB}")
+ list (REMOVE_DUPLICATES PETSC_EXTERNAL_LIB)
+ string (REPLACE ";" " " PETSC_EXTERNAL_LIB "${PETSC_EXTERNAL_LIB}")
++string (REPLACE "-lhdf5_hl -lhdf5" "-lhdf5hl_fortran -lhdf5_fortran -lhdf5_hl -lhdf5" PETSC_EXTERNAL_LIB "${PETSC_EXTERNAL_LIB}")
```
Comment (bernhardkaindl) [inline var/spack/repos/builtin/packages/damask-grid/PETSc-HDF5-Fortran.patch]: I've got this when building `spack install --fail-fast --test=root -v damask`:
```
  >> 662    /usr/bin/ld: cannot find -lhdf5hl_fortran
  >> 663    collect2: error: ld returned 1 exit status
     664    src/CMakeFiles/DAMASK_mesh.dir/build.make:954: recipe for target 'src/DAMASK_mesh' failed
```
Assumption: As you are linking with `-lhdf5hl_fortran`, you have to ensure that spack selects the `hl` variant of `hdl5` (which is not enabled by default) as package to link damask with.
```diff
--- a/var/spack/repos/builtin/packages/damask-grid/package.py
+++ b/var/spack/repos/builtin/packages/damask-grid/package.py
@@ -21 +21 @@ class DamaskGrid(CMakePackage):
-    depends_on('hdf5+fortran')
+    depends_on('hdf5+fortran+hl')
diff --git a/var/spack/repos/builtin/packages/damask-mesh/package.py b/var/spack/repos/builtin/packages/damask-mesh/package.py
index 955bdbc7dc..87104c8a3f 100644
--- a/var/spack/repos/builtin/packages/damask-mesh/package.py
+++ b/var/spack/repos/builtin/packages/damask-mesh/package.py
@@ -21 +21 @@ class DamaskMesh(CMakePackage):
-    depends_on('hdf5+fortran')
+    depends_on('hdf5+fortran+hl')
```
And, there is a "_hl" not jusr "hl":
```suggestion
+string (REPLACE "-lhdf5_hl -lhdf5" "-lhdf5_hl_fortran -lhdf5_fortran -lhdf5_hl -lhdf5" PETSC_EXTERNAL_LIB "${PETSC_EXTERNAL_LIB}")
```
And you'd have to update the sha256 of the patch after the fix.
Easier to change is:
```py
   def patch(self):
        filter_file('";" " "',
                    '"-lhdf5_hl -lhdf5" "-lhdf5_hl_fortran -lhdf5_fortran -lhdf5_hl -lhdf5"'
                    'CMakeLists.txt')
```
This fixes the build for me. I'll try to push it to your branch.

Comment (MarDiehl) [reply var/spack/repos/builtin/packages/damask-grid/PETSc-HDF5-Fortran.patch]: thanks and sorry for the sloppy update.
Let me know if something is still broken, I will then take care of this.

Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/damask-grid/PETSc-HDF5-Fortran.patch]: Thanks and sorry for stepping ahead on this one. I'll just notify you in the future.

Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/damask-grid/PETSc-HDF5-Fortran.patch]: I'll leave it to you to fix this properly:

I don't think that it is correct to add a second, different variant of `hdf5` into the build: `petsc` depends on `hdf5+mpi` while damask adds `hdf5+fortran+hl`. One library, differently built used for `damask` and `petsc` does not look good to me.

IMHO, you'd have to add a variant to `petsc`'d package.py which depends on `hdf5+fortran+hl` and remove the depends_on for hdf5 from damask, and reuse the hdf5 used by `petsc` to do this properly.

**Comment by bernhardkaindl** [comment]
Foreword:

In case hdf5 with fortran+hl is not actually needed by damask and it only was a workaround on your side to add missing flags which weren't automatically retrieved from petsc's pkg-config file petsc.pc because maybe pkg-config isn't installed on your system(s) either, then the hdf5 variants can be removed from this version (which would be preferable).

The remaining build fix is then to ensure to have depends_on for pkgconfig and fftw, and add the fftw3 link flags to the CMakeLists.txt file.

Commit message of this push, I hope it helps to understand the changes:

fix the build, use variant of hdf5 to get hdf5+mpi+fortran+hl

This push is rebased to the latest develop branch: Needed to prevent
the attempt to compile llvm@13 with older compilers which fails
when trying to satisfy the llvm dependency of mesa(OpenGL) for vtk,
which is a dependency of py-damask.

Description of the build fix for the hdf5_hl_fortran link issue:

The CMakeLists.txt uses:
pkg_check_modules (PETSC REQUIRED PETSc>=3.12.0 PETSc<3.17.0)

This means that cmake requires pkg-config to find petsc's compiler and
linker flags. If it isn't provided (like in my Ubuntu 18.04 container),
it won't get the needed flags from petsc's lib/pkgconfig/petsc.pc file.

Add pkgconfig as dependency to ensure it is always present when building.

Among many other flags which migth be important too, it adds these flags
from the default hdf5 variant:

-lHYPRE -lsuperlu_dist -lopenblas -lhdf5_hl -lhdf5 -lparmetis -lmetis
-lm -lz -lstdc++ -ldl -lmpi_usempif08 -lmpi_usempi_ignore_tkr
-lmpi_mpifh -lmpi -lgfortran -lm -lgfortran -lm -lgcc_s -lquadmath
-lpthread -lquadmath -lstdc++ -ldl

With the new hdf5 variants added and requested for/by petsc in this
commit, the flags from lib/pkgconfig/petsc.pc are:

-lHYPRE -lsuperlu_dist -lopenblas -lhdf5_hl_fortran -lhdf5_hl_f90cstub
-lhdf5_hl -lhdf5_fortran -lhdf5_f90cstub -lhdf5 -lparmetis -lmetis -lm
-lz -lstdc++ -ldl -lmpi_usempif08 -lmpi_usempi_ignore_tkr -lmpi_mpifh
-lmpi -lgfortran -lm -lgfortran -lm -lgcc_s -lquadmath -lpthread
-lquadmath -lstdc++ -ldl

This adds the flags related to -lhdf5_hl_fortran.

The remaining bit is then a problem where damask uses fftw3 in
src/grid/discretization_grid.f90 and src/grid/spectral_utilities.f90
does not link with it and I could not find linker flags for it.

Ideally, this should be done with pkg_check_modules() like for petsc,
but it appears that fftw3 does not yet deliver an fftw3mpi.pc file for
providing the flags for fftw3+mpi. Thankfully the flags are simple for
now, so they can be patched after the -lz flag in CMakeLists.txt.
This is not a clean solution, so it should be replaced by solution
provided by the damask's CMakeLists.txt itself, where it should ensure
the presence of the fftw3 library and and the linker flags for it,
but preferably working with the upstream fftw3 project to get the
linker flags for fftw3+mpi thru pkg-config instead.

**Comment by MarDiehl** [comment]
@bernhardkaindl: Many thanks for your help and the detailed explanation. I've reinstalled spack and all its packages which took a while. Now there is a working solution (https://github.com/MarDiehl/spack/tree/damask-3.0.0-alpha5-second-try).

In the end, it is only required to link additionally to `-lhdf5_fortran`. The `hl` (high-level) variants are not needed.

That leaves me with two questions
1) Should I apply these changes to this branch via a force commit to keep the history clean?
2) There are two options for FFTW: require `petsc+fftw` (add linker options automatically, might result in installing two PETSc versions) or add FFTW as a separate dependency (for `damask-grid` only) and patch the linker command.

**Comment by bernhardkaindl** [comment]
> @bernhardkaindl: Many thanks for your help and the detailed explanation. I've reinstalled spack and all its packages which took a while. Now there is a working solution (https://github.com/MarDiehl/spack/tree/damask-3.0.0-alpha5-second-try).
> 
> In the end, it is only required to link additionally to `-lhdf5_fortran`. The `hl` (high-level) variants are not needed.
> 
> That leaves me with two questions
> 1. Should I apply these changes to this branch via a force commit to keep the history clean?

Yes, you can force-push your branch/commit(s) here to remove intermediate commits which are obsolete.

> 2. There are two options for FFTW: require `petsc+fftw` (add linker options automatically, might result in installing two PETSc versions) or add FFTW as a separate dependency (for `damask-grid` only) and patch the linker command.

Either one would be ok for me as long as
* your build works for you and
* I can also confirm that it builds for me (that would be my "wish", another reviewer might have other wishes)

My last push to this branch was to add fftw as a direct dependency with `-lfftw3 -lfftw3_mpi`.
Not the cleanest way, I guess, I can only say it compiled.

I kept `petsc+fftw` as well. I guess you ment that using it might result in two `fftw` versions built (not PETSc versions), right?

I guess I have now understood what you might refer to:

Now I think I've made an interesting discovery checking that the dependency  of `petsc+fftw` actually does:
```py
grep -n fftw var/spack/repos/builtin/packages/petsc/package.py
126:    variant('fftw', default=False,
174:    conflicts('+fftw', when='~mpi', msg=mpi_msg)
295:    depends_on('fftw+mpi', when='+fftw+mpi')
420:                ('fftw:mpi', 'fftw', True, True),
```
Adding `petsc+fftw` does not cause a `depends_on('fftw')` added to petsc, because as `petsc+mpi` is a default of `petsc`, the effective dependency added is `fftw+mpi` (line 295). Then again the default variant of `fftw` is actually `fftw+mpi`.

While that would be two builds of `fftw`, both are currently `fftw+mpi`. At least not different, incompatible builds.

In my push, I removed the `+fftw` from the `depends_on('petsc')`: https://github.com/spack/spack/blob/883be4f2a637e92ecb062e80e0f0bf788a2694f9/var/spack/repos/builtin/packages/damask-mesh/package.py#L18
https://github.com/spack/spack/blob/883be4f2a637e92ecb062e80e0f0bf788a2694f9/var/spack/repos/builtin/packages/damask-mesh/package.py#L22

**Comment by bernhardkaindl** [comment]
### To be sure that we've only one build of `fftw` (ensure no conflict of different versions):

If the `PETSc` that you need doesn't need itself `+fftw`, then I'd propose these dependencies:
```
depends_on('fftw+mpi') 
depends_on('petsc~fftw') # with versions and when= added as needed
```
With this, it should be ensured that the only `fftw` dependency is only for `damask-grid` itself, not for `PETSc`. For now, you'd have to add the needed -lfftw* flags as needed to the linker command.

I'd prefer to have them added like I pushed by extending the `-lz` in the linker line using filter_file instead of a patch because the package.py shows what it does and it is easy to change:
```
    def patch(self):
        filter_file(' -lz ', ' -lz -lfftw3 -lfftw3_mpi', 'CMakeLists.txt')
```
If possible instead, according to https://stackoverflow.com/questions/44284275/passing-compiler-options-cmake it should be possible to export LDFLAGS in the environment and cmake should add that to it's linker flags.

I am not sure if I tried that, but you can set the environment for the `cmake` build like this (fish is also a `CMakePackage`):
```py
grep -rB1 LDFL var/spack/repos/builtin/packages/fish/package.py
    def setup_build_environment(self, env):
        env.append_flags('LDFLAGS', self.spec['ncurses'].libs.link_flags)
```
In this example, fish reads the `libs` propery of the `ncurses` dependency and `.link_flags` prepends them with `-l`, turning those into `-lncurses` and `-lncursesw` and the same for termlib:
```py
grep -rnhA6 -B1 libs\( var/spack/repos/builtin/packages/ncurses/
160-    @property
161:    def libs(self):
162-        libraries = ['libncurses', 'libncursesw']
163-
164-        if '+termlib' in self.spec:
165-            libraries += ['libtinfo', 'libtinfow']
166-
167-        return find_libraries(libraries, root=self.prefix, recursive=True)
```

**Comment by MarDiehl** [comment]
Different builds
==========
> I kept `petsc+fftw` as well. I guess you ment that using it might result in two fftw versions built (not PETSc versions), right?

I was indeed referring to PETSc, but _versions_ was not precise. I believe in spack terminology different builds are called _concretizations_. At least I ended up having PETSc with two different hashes when installing `damask-grid` and `damask-mesh` separately.
PETSc's build system is like a small package manager in itself, it is possible to install different configurations of PETSc that rely on various external packages. The packages can be downloaded (e.g. `--download-hdf5`) if existing packages (`--with-hdf5`) are not available. Upstream PETSc handles that via `PETSC_ARCH`, on spack, which uses only existing packages, it results on two different concretizations/builds/hashes.

PETSc as a package manager
==================
For manual DAMASK (`damask-grid`/`damask-mesh`) installations, using PETSc as a 'package manager' is the fail-safe option: `damask-grid` needs FFTW, so PETSc should be installed with `--download-fftw` or `--with-fftw`. This ensures that all required libraries are correctly installed and PETSc can take care of assembling the linker command. The spack package was using this approach until PETSc 3.16 which stopped linking to the Fortran bindings of HDF5.

For spack, this approach is not convenient because FFTW is not a default dependency (probably because PETSc does not use it itself) and because PETSc does not provide the linker options for Fortran (because PETSc uses only the C bindings of HDF5). To get a suitable PETSc concretization, it would need to be build with FFTW  and with `--download-hdf5-fortran-bindings=1`. However, since we can make sure that FFTW and HDF5 with Fortran bindings are installed via `depends`, we can also modify the linker options directly.

MPI variants
========
That `petsc+fftw` results in `fftw+mpi` is sensible for `damask-grid`. The same holds for HDF5, which is also required to have MPI support for PETSc.

Patch vs LDFLAGS
============
```
    def setup_build_environment(self, env):
        env.append_flags('LDFLAGS', self.spec['hdf5'].libs.link_flags)
```
seems to work


```
    def setup_build_environment(self, env):
        env.append_flags('LDFLAGS', self.spec['fftw'].libs.link_flags)
```
does not work (linker command is incomplete)

TLDR
===
Linker flags for FFTW and a dependency to `fftw+mpi` are now added manually, this allows to use a default PETSc installation (without FFTW) and follows the same procedure as for HDF5 Fortran bindings. Changes are force pushed to the PR.

**Review by bernhardkaindl — APPROVED** [review]
Thanks for the explanation: Yes I agree, and it's a good improvement for the build too.

@MarDiehl I'll open a new PR with you as reviewer to discuss and solve my remaining findings and ideas.

**Comment by MarDiehl** [comment]
> Thanks for the explanation: Yes I agree, and it's a good improvement for the build too.
> 
> @MarDiehl I'll open a new PR with you as reviewer to discuss and solve my remaining findings and ideas.

sure!
